source/common/ssl/context_manager_impl.cc

```diff
@@ -22,23 +58,77 @@ void ContextManagerImpl::releaseContext(Context* context) {
 
 ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
                                                             ClientContextConfig& config) {
-
   ClientContextPtr context(new ClientContextImpl(*this, scope, config));
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
   return context;
 }
 
-ServerContextPtr ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                            ServerContextConfig& config) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, runtime_));
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+ServerContextPtr ContextManagerImpl::createSslServerContext(
+    const std::string& listener_name, const std::vector<std::string>& server_names,
+    Stats::Scope& scope, ServerContextConfig& config, bool skip_context_update) {
+  ServerContextPtr context(new ServerContextImpl(*this, listener_name, server_names, scope, config,
+                                                 skip_context_update, runtime_));
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
+
+  // Save mappings.
+  if (server_names.empty()) {
+    map_exact_[listener_name][EMPTY_STRING] = context.get();
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        map_wildcard_[listener_name][name] = context.get();
+      } else {
+        map_exact_[listener_name][name] = context.get();
+      }
+    }
+  }
+
   return context;
 }
 
+ServerContext* ContextManagerImpl::findSslServerContext(const std::string& listener_name,
+                                                        const std::string& server_name) {
+  // Find Ssl::ServerContext to use. The algorithm for "www.example.com" is as follows:
+  // 1. Try exact match on domain, i.e. "www.example.com"
+  // 2. Try exact match on wildcard, i.e. "*.example.com"
+  // 3. Try "no SNI" match, i.e. ""
+  // 4. Return no context and reject connection.
+
+  std::shared_lock<std::shared_timed_mutex> lock(contexts_lock_);
```

Comment:
can you throw a TODO here (attributed to me if you want) about removing locking from this function. With R/W lock it's fine but would prefer to ultimately make it lockless.

Reply:
Done. I think we could fix that by using `std::shared_ptr`, but I didn't want to block this PR on it. Also, keep in mind that whatever we're going to do now, it's most likely going to change with SDS.